### DISCUSSION TITLE
Docking port - improvements!

### DISCUSSION THREAD

**Discussion by MaaiG** [FakeFishGames/Barotrauma#10198] [Ideas and suggestions]
**1. Multiple docking - submarine docking multiple docking ports.**
We can’t do this now. The submarine or shuttle can dock only with one `docking port`. Unfortunately, docking at once with two ports is not possible. You can only change or switch active docking ports, but you cannot use multiple docking ports at once.
This state of affairs greatly limits the design solutions for the design of underwater vehicles of this type: ![Image 002_13-01_3v2](https://user-images.githubusercontent.com/93348499/199218171-de968198-664f-4576-b646-056bff29026a.png)
**2. Add the settings "Do not display on sonar" to the object `Docking Port` properties.**
On the complex submarine with several shuttles and escape capsules, the sonar screen appears littered (too many green lines).
In fact, to facilitate navigation,  the player only uses the main docking port and the display of others should be optional.
The lack of a green line on sonar does not affect the ability to dock - such ports can be docked visually or automatically.

![199213533-1ea3ca17-2674-4696-b29a-387fda209313 (1) (1)](https://user-images.githubusercontent.com/93348499/199220526-6f2c952b-1c14-4712-8b2e-789028f701a9.png)

**3. Add `signal_in` and `signal_out`.** https://github.com/Regalis11/Barotrauma/discussions/8146
You can control the devices of docked objects or receive their data through its docking port. Determine the type of docked object, etc... Anything you want. It is enough to have one group of connectors "signal_in" and "signal_out".![199217978-ae44bf61-48a1-49ba-8dcd-23c0ebfb4e02](https://user-images.githubusercontent.com/93348499/199220644-4355ec61-0bad-4fda-9560-5cf9952bb13d.png)

**Comment by MapleWheels** [comment]
Everything here is great except the actual multi-docking idea. The problem is that you are assuming that the docks would be lined up. When you dock, the game auto-aligns everything to the dock when it connects. What happens if one of the docks is "misaligned"?

**Reply by MaaiG** [reply]
> What happens if one of the docks is "misaligned"?

When designing my submarines, I carefully ensure that all docking ports, stairs -  have a precise connection. In my designs I can guarantee the "docking system standards" and adhere to them.
If it is not aligned then it is not designed to dock these objects. It will be ignored.

By default, the game can continue to dock only one port.
The sub-builders just needs ability to be able to dock several ports simultaneously. The command on this docking will exist through the logical circuit of the sub-builder. In addition, the main docking port may have a "primary" mark while the additional port will be in the "secondary" position (by default).
There’s no problem docking aligned docking ports.

Sub-builders  can make any schemes. The main thing is to give him the opportunity. Vanilla game may not create such game locations, but sub-builders will be able to make their own.
**The main purpose of this proposal is to implement custom submarines and their shuttles, lifeboats, lifts, weapons or other parts. That is, for a closed system! But it can be used in the vanilla game.**

It is necessary to systematize and regulate such things as docking. There should be recommendations for setting up docking ports for sub-builders.
Even in vanilla submarines there are docking ports and stairs with a displacement of several pixels.

For a sub-builders. there should be recommendations:
- Minimum docking port height above submarine hull
- Location of the connecting staircase
- Maybe a label of the docking port "Primary" - for docking with standard vanilla outposts, etc.
- Etc...  Down to "Docking System Standard"!

In addition, you can always configure the ID system for the docking port.
Proximity sensor in docking port sends its ID to wifi component
The incoming docking port receives this ID and, if it matches, sends a command to dock. If the ID does not match, no docking permissions will be issued. The submarine won’t be able to dock its secondary docking port.
_With such things you can think of a lot of game moments and situations, up to password guessing to dock to the object )) or cracking the security system, password interception when docking another ship and other electronic game radio and "Electronic warfare", etc._
